lib/src/main/kotlin/StatelessSolver.kt

```diff
@@ -15,12 +15,22 @@ class StatelessSolver<TState, TAction>(
 
     private var root : StateActionNode<TAction>? = null
 
-    fun buildTree() {
+    fun buildTree(): MutableList<Double> {
         initialize()
 
-        for (i in 0 until iterations) {
+        var rewardTracker = mutableListOf<Double>()
```

Comment:
Hmm I don't like this implementation, it's not extensible. However, I do recognize the need to be able to keep track of and potentially alter these parameters. I think it will be an extension point during the backprop step.

Reply:
@JunTaoLuo im currently unable to do this unless i heavily modify the `State{less/ful}Solver` because the variables, namely `root` and `calculateUCT` class methods are `private`, if we change their visibility to `open protected` then their subclasses should get access to those variables. The code below should work if we make this edit codebase wise 

```
package MDPSolver

import MDP
import kotlin.random.Random
import StatelessSolver
import kotlin.math.ln
import kotlin.math.sqrt

class StatelessSolverTracked<TState, TAction>(
        private val mdp: MDP<TState, TAction>,
        private val random: Random,
        private val iterations: Int,
        private val simulationDepthLimit: Int,
        private val explorationConstant: Double,
        private val rewardDiscountFactor: Double,
        private val verbose: Boolean): StatelessSolver<TState, TAction>(
        mdp,
        random,
        iterations,
        simulationDepthLimit,
        explorationConstant,
        rewardDiscountFactor,
        verbose) {


    fun buildTreTracked(): MutableList<Double> {
        initialize()

        var rewardTracker = mutableListOf<Double>()

        for (i in 0..iterations) {
            iterateStep()

            // println(root!!.children.toString())
            // println(root!!.children.maxByOrNull { a -> calculateUCT(a)} )
            var ns = root!!.children.maxByOrNull { a -> calculateUCT(a)}!!.n
            var explorationFactor = explorationConstant*sqrt(ln(i.toDouble())/ns)
            rewardTracker.add(explorationFactor)

        }
        return rewardTracker
    }
    
}
```
Otherwise, should we keep the original implementation, and use some flag to trigger the `rewardTracker` tracking?